client/driver/rkt.go

```diff
@@ -435,66 +472,181 @@ func (d *RktDriver) Start(ctx *ExecContext, task *structs.Task) (*StartResponse,
-			prepareArgs = append(prepareArgs, fmt.Sprintf("--mount=volume=%s,target=%s", volName, parts[1]))
 		}
-	}
 
-	// Inject environment variables
-	for k, v := range ctx.TaskEnv.Map() {
-		prepareArgs = append(prepareArgs, fmt.Sprintf("--set-env=%s=%s", k, v))
-	}
+		// Inject environment variables
+		for k, v := range ctx.TaskEnv.Map() {
+			prepareArgs = append(prepareArgs, fmt.Sprintf("--set-env=%s=%s", k, v))
+		}
 
-	// Image is set here, because the commands that follow apply to it
-	prepareArgs = append(prepareArgs, img)
+		// Image is set here, because the commands that follow apply to it
+		prepareArgs = append(prepareArgs, img)
 
-	// Check if the user has overridden the exec command.
-	if driverConfig.Command != "" {
-		prepareArgs = append(prepareArgs, fmt.Sprintf("--exec=%v", driverConfig.Command))
-	}
+		// Check if the user has overridden the exec command.
+		if driverConfig.Command != "" {
+			prepareArgs = append(prepareArgs, fmt.Sprintf("--exec=%v", driverConfig.Command))
+		}
+
+		// Add memory isolator
+		prepareArgs = append(prepareArgs, fmt.Sprintf("--memory=%vM", int64(task.Resources.MemoryMB)))
 
-	// Add memory isolator
-	prepareArgs = append(prepareArgs, fmt.Sprintf("--memory=%vM", int64(task.Resources.MemoryMB)))
+		// Add CPU isolator
+		prepareArgs = append(prepareArgs, fmt.Sprintf("--cpu=%vm", int64(task.Resources.CPU)))
+
+	} else {
+		allocACName, err := appctypes.NewACName(allocVolName)
+		if err != nil {
+			return nil, fmt.Errorf("[ERR] driver.rkt: failed to set ACName: %s", err)
+		}
 
-	// Add CPU isolator
-	prepareArgs = append(prepareArgs, fmt.Sprintf("--cpu=%vm", int64(task.Resources.CPU)))
+		localACName, err := appctypes.NewACName(localVolName)
+		if err != nil {
+			return nil, fmt.Errorf("[ERR] driver.rkt: failed to set ACName: %s", err)
+		}
+
+		secretsACName, err := appctypes.NewACName(secretsVolName)
+		if err != nil {
+			return nil, fmt.Errorf("[ERR] driver.rkt: failed to set ACName: %s", err)
+		}
+
+		podManifest.Volumes = []appctypes.Volume{
+			{
+				Name:   *allocACName,
+				Kind:   "host",
+				Source: ctx.TaskDir.SharedAllocDir,
+			},
+			{
+				Name:   *localACName,
+				Kind:   "host",
+				Source: ctx.TaskDir.SharedAllocDir,
+			},
+			{
+				Name:   *secretsACName,
+				Kind:   "host",
+				Source: ctx.TaskDir.SharedAllocDir,
+			},
+		}
+
+		mounts := []appcschema.Mount{
+			{
+				Volume: *allocACName,
+				Path:   allocdir.SharedAllocContainerPath,
+			},
+			{
+				Volume: *localACName,
+				Path:   allocdir.TaskLocalContainerPath,
+			},
+			{
+				Volume: *secretsACName,
+				Path:   allocdir.TaskSecretsContainerPath,
+			},
+		}
+
+		var environment appctypes.Environment
+		for name, value := range ctx.TaskEnv.Map() {
+			environment = append(environment, appctypes.EnvironmentVariable{
+				Name:  name,
+				Value: value,
+			})
+		}
+
+		fetchArgs := make([]string, 0, 50)
```

Comment:
The `fetchArgs` variable could be created inside the `fetchImage` function. So `fetchImage` could just take globalArgs. The reason is that `fetchImage` knows exactly how large the array can be, so you could avoid using some magic number like 50, and `fetchImage` can add the `"fetch"` arg itself.

Reply:
Another reason is that `fetchImage` modifies `fetchArgs` so it gets changed after each iteration. And I don't think that `append` function makes a deepcopy of the modified slice. That looks fishy.

Reply:
My understanding is that `fetchImage` gets a new reference to the newly created slice while the line:

```
ID, err := d.fetchImage(app.Image, fetchArgs)
```

always refers to the `fetchArgs` created outside the for loop. As a result the state of the `fetchArgs` remains unaltered after each iteration.

Here's a snippet for validation: https://play.golang.org/p/0kq6tCLGk6d

However I agree that we can declare `fetchArgs` inside the function itself and get rid of the magic number.

Reply:
The slices are a bit tricky and I don't remember all the details. So in this case it seems to work fine as long as we don't modify the contents that were already in the slice. Consider this: https://play.golang.org/p/xvLJbVqdd48